### PR TITLE
Introduce crewai tracing

### PR DESCRIPTION
**Pull request by TomeHirata**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/TomeHirata/mlflow/pull/13903?quickstart=1)

#### Install mlflow from this PR

```
# Use `%sh` to run this command on Databricks
OPTIONS=$(if pip freeze | grep -q 'mlflow @ git+https://github.com/mlflow/mlflow.git'; then echo '--force-reinstall --no-deps'; fi)
pip install $OPTIONS git+https://github.com/mlflow/mlflow.git@refs/pull/13903/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13903
```

</p>
</details>

### Related Issues/PRs

 #xxx
N/A

#### Context
CrewAI is a popular framework for orchestrating AI agents with more than 20k stars on Github as of Nov 26, 2024. Because of the popularity of the framework, we would like to integrate mlflow tracing with CrewAI.

#### What is the change
This PR introduces an auto-tracing integration with CrewAI. Users can enable the auto-tracing with `mlflow.crewai.autolog()` and see the tracing on the mlflow tracking server, like the picture below.

Screenshot 2024-11-28 at 18 49 24

#### Supported versions
v0.80.0 ~ v0.83.0

#### Known limitation
Due to some technical challenges in the current safe patch logic, asynchronous methods are not covered by auto-tracing now.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [x] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/crewai/__init__.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        disable: If ``True``, disables the CrewAI autologging. Default to ``False``.
+        silent: If ``True``, suppress all event logs and warnings from MLflow during CrewAI
+            autologging. If ``False``, show all events and warnings.
+    """
+    # TODO: handle asynchronous tasks and crew executions
+    # TODO: interface of tool in CrewAI is changing drastically. Add patching once it's stabilized
+    import crewai
+
+    safe_patch(
+        FLAVOR_NAME,
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:37]: I would define a map like this for better readability:
```
class_to_methods_to_patch = {
   crewai.Crew: ["kickoff", "kioff_for_each", "train"],
   ...
   crewai. ShortTermMemory: ["save", "search", "reset"],
}

for module, methods_to_patch in class_to_methods_to_patch.items():
     try:
         safe_patch(FLAVOR_NAME, module, methods_to_patch, patched_class_call)
     except...
```

File: mlflow/crewai/__init__.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+@experimental
+@autologging_integration(FLAVOR_NAME)
+def autolog(
+    log_traces: bool = True,
+    disable: bool = False,
+    silent: bool = False,
+):
+    """
+    Enables (or disables) and configures autologging from CrewAI to MLflow.
+    Only synchronous calls are supported. Asynchnorous APIs and streaming are not recorded.
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:23]: Q: Is streaming available in CrewAI?

File: mlflow/crewai/__init__.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.UserMemory,
+            "search",
+            patched_class_call,
+        )
+
+        safe_patch(
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:142]: Is there any limit (either theoretical or practical) for number of entities saved to `EntityMemory` in one task? If it can go sth like 100>, we probably want to either not patching or organize them into a parent span, otherwise the span UI becomes hart to navigate.

File: mlflow/crewai/autolog.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+            return result
+
+
+def _get_span_type(cls, task_name: str) -> str:
+    from crewai.flow.flow import Flow
+
+    if issubclass(cls, Flow):
+        return SpanType.CHAIN
+    span_type_mapping = {
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:36]: Looks like we don't need method level mapping here? Each class maps to single span type (probably our span types aligns with their class design). We MAY need them but seems YAGNI at this moment.

File: mlflow/crewai/autolog.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        "Crew.kickoff": SpanType.CHAIN,
+        "Crew.kickoff_for_each": SpanType.CHAIN,
+        "Crew.train": SpanType.CHAIN,
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:39]: nit: It's subjective but `AGENT` seems to be more appropriate type for `Crew` than `CHAIN`. For "chain" I would imagine a static DAG like LangChain (at least this was the origin of the type name), while `Crew` is a dynamic execution by multiple agents.

File: mlflow/crewai/__init__.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        "call",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Flow,
+        "kickoff",
+        patched_class_call,
+    )
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:83]: As discussed in the design doc, let's patch `BaseTool` class to capture tool execution information.

File: mlflow/crewai/__init__.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.LongTermMemory,
+            "search",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.LongTermMemory,
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:123]: Is `reset` called during agent execution? If it is stand-alone operation that users explicitly does outside agent, I don't think we need to patch it?

File: mlflow/crewai/autolog.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    }
+    return span_type_mapping.get(task_name, SpanType.UNKNOWN)
+
+
+def _is_serializable(value):
+    try:
+        json.dumps(value, cls=TraceJSONEncoder, ensure_ascii=False)
+        return True
+    except (TypeError, ValueError):
+        return False
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:65]: Is there any way we can check this (even in exchange to some cost of accuracy) without actually serializing the object? For context, the serialization actually has visible overhead right now for large spans e.g. document retrieval.

File: mlflow/crewai/autolog.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        k: v.__dict__ if hasattr(v, "__dict__") else v
+        for k, v in arguments.items()
+        if v is not None and _is_serializable(v)
+    }
+
+
+def _set_span_attributes(span: LiveSpan, instance):
+    instance_name = instance.__class__.__name__
+    ## memory does not have helpful attributes
+    if instance_name == "Crew":
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:87]: Can we use `isinstance` here? Class name comparison doesn't handle subclasses.

File: mlflow/crewai/autolog.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        for key, value in instance.__dict__.items():
+            key = f"crewai.crew.{key}"
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:89]: We don't use any prefix for attributes in other flavors. Can we remove them for consistency?

File: mlflow/crewai/autolog.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+    if config.log_traces:
+        fullname = f"{self.__class__.__name__}.{original.__name__}"
+        with mlflow.start_span(
+            name=fullname,
+            span_type=_get_span_type(self.__class__, fullname),
+        ) as span:
+            inputs = _construct_full_inputs(original, self, *args, **kwargs)
+            span.set_inputs(inputs)
+            _set_span_attributes(span=span, instance=self)
```
Comment (B-Step62) [inline mlflow/crewai/autolog.py:22]: Let's wrap this function within try-catch. In case any exception happens e.g. `AttributeError`, we don't want to drop the span entirely but instead gracefully skip that attributes.

File: mlflow/ml-package-versions.yml [hunk 12]
```diff
@@ -75,7 +75,7 @@ keras:
 
   models:
     minimum: "3.0.2"
-    maximum: "3.6.0"
+    maximum: "3.7.0"
```
Comment (B-Step62) [inline mlflow/ml-package-versions.yml:78]: Let's revert unrelated changes here. See https://github.com/mlflow/mlflow/pull/13851#discussion_r1860827148 for how to prevent auto-update.

File: tests/crewai/test_crewai_autolog.py
Comment (B-Step62) [inline tests/crewai/test_crewai_autolog.py:1]: Thanks for the great test coverage!😁

File: mlflow/crewai/__init__.py (hunk 2 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:23]: To my understanding, it's not supported.

File: mlflow/crewai/__init__.py (hunk 6 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:83]: As I added the comment, the BaseTool class is actually not used by CrewAI. Paching BaseTool.run does not work.

File: mlflow/crewai/__init__.py (hunk 7 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:123]: I don't think it's something users explicitly run?

File: mlflow/crewai/__init__.py (hunk 6 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:83]: It is used tho  https://docs.crewai.com/concepts/tools

Comment (B-Step62) [reply mlflow/crewai/__init__.py:83]: btw for built-in tools are defined in https://github.com/crewAIInc/crewAI-tools not the core repo, which implement BaseTool

File: mlflow/crewai/__init__.py (hunk 7 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:123]: Do you know when it is used?

File: mlflow/crewai/__init__.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:23]: yea, can we remove it from thd docstring?

File: mlflow/crewai/__init__.py (hunk 3 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:142]: There are only 4 types of memories and agent just call several methods for each, I don't think it will go that many.

File: mlflow/crewai/autolog.py (hunk 5 above)
Comment (TomeHirata) [reply mlflow/crewai/autolog.py:39]: I think it's actually confusing to mark crew as agent, since it's orchestralation layer, and the actual procedure of crew is similar to DAG execution. Then, I would introduce a new type instead of reusing agent if Chain does not match.

File: mlflow/crewai/__init__.py (hunk 6 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:83]: I mean, the function itself is not used directly, inside the library which prevent patching. Also the implementation of Tool is changing, even the latest code is very different from the master branch. I'd wait until having more stability.

File: mlflow/crewai/__init__.py (hunk 7 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:123]: I basically patched the public interface of memory classes, need to dig if we really want to know what exactly logic triggers reset.

File: mlflow/crewai/autolog.py (hunk 8 above)
Comment (TomeHirata) [reply mlflow/crewai/autolog.py:65]: We could check acceptable classes of the encoder first and then also check the references of the pydamic model. But if you care the latency due to a large responce from llm, I didn't call this logic for outputs.
https://docs.pydantic.dev/latest/concepts/forward_annotations/#cyclic-references

File: mlflow/crewai/autolog.py (hunk 4 above)
Comment (TomeHirata) [reply mlflow/crewai/autolog.py:36]: Let me change it to class based approach here.

File: mlflow/crewai/autolog.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/crewai/autolog.py:39]: Got it, yea if it is DAG-like execution, let's go with CHAIN🙂

File: mlflow/crewai/__init__.py (hunk 3 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:142]: You're right, but is there any better way to make this more readable? We shouldn't change the span structure, but there is nothing bad in adding a parent span for a real function call or changing span names.

Screenshot 2024-12-02 at 9 38 03

File: mlflow/crewai/__init__.py (hunk 6 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:83]: I did a bit of research:
* In version `0.83`, they are using LangChain's tool implementation and call `tool.invoke` within the agent. https://github.com/crewAIInc/crewAI/blob/v0.83.0/src/crewai/agent.py#L402
* In the latest `main` branch, they've added `CrewStructuredTool` to remove LangChain dependency.

The new structured tool class will be a standard for tool calling. I agree the code base is changing rapidly right now, but we must have tool spans as well, otherwise users cannot debug/measure tool execution. Can we leave a TODO comment and file a time-bounded JIRA?

File: mlflow/crewai/autolog.py (hunk 8 above)
Comment (B-Step62) [reply mlflow/crewai/autolog.py:65]: > But if you care the latency due to a large responce from llm, I didn't call this logic for outputs.

Interesting. Would you elaborate why we only need this for input and what happens if we remove this?

Comment (B-Step62) [reply mlflow/crewai/autolog.py:65]: Btw I faced this warning when running the code on my notebook. Is this perhaps related to this logic?
```
2024/12/02 01:23:52 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/databricks/python/lib/python3.12/site-packages/pydantic/functional_validators.py:659: UserWarning: Pydantic serializer warnings:
  Expected `InstanceOf` but got `CrewAgentExecutor` - serialized value may not be as expected"
```

Comment (TomeHirata) [reply mlflow/crewai/autolog.py:65]: The input object has some attributes that have cyclic references, and the serialization fails because of the cyclic references. We could add a validation if the class is defined by us like the code in the pydantic documentation, but since this class is defined by crewai library, the safest way is trying to serialize, like my code here. 
https://docs.pydantic.dev/latest/concepts/forward_annotations/#self-referencing-or-recursive-models

File: mlflow/crewai/__init__.py (hunk 6 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:83]: I've added a comment [here](https://github.com/mlflow/mlflow/pull/13903/files/2c9d3d3f7729f425e841b5e662091c4a0756bbf9#diff-c754f6b0a60b7373d437491941f830ce9a6d616e26bd39366b0d984b1d3808b1R33), let me create a ticket for it too.

File: mlflow/crewai/autolog.py (hunk 8 above)
Comment (TomeHirata) [reply mlflow/crewai/autolog.py:65]: The warning occurs due to the mismatch between the expected class and the actual object inside the crewai.Agent implementation. It is displayed whenever the agent object is serialized.

File: mlflow/crewai/__init__.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    import crewai
+
+    try:
+        class_method_map = {
+            crewai.Crew: ["kickoff", "kickoff_for_each", "train"],
+            crewai.Agent: ["execute_task"],
+            crewai.Task: ["execute_sync"],
+            crewai.LLM: ["call"],
+            crewai.Flow: ["kickoff"],
+            crewai.agents.agent_builder.base_agent_executor_mixin.CrewAgentExecutorMixin: [
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:50]: Perhaps this patching is not taking effect? I've run the example but this span was not created🤔

File: mlflow/crewai/__init__.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        for cls, methods in class_method_map.items():
+            for method in methods:
+                safe_patch(
+                    FLAVOR_NAME,
+                    cls,
+                    method,
+                    patched_class_call,
+                )
+    except AttributeError as e:
+        _logger.error("An exception happens when applying auto-tracing to crewai. Exception: %s", e)
```
Comment (B-Step62) [inline mlflow/crewai/__init__.py:75]: Shall we try-catch each method patching independently? E.g., when a class `Knowledge` is renamed, should we stop tracing at all or still trace other classes?

**Review by B-Step62 — APPROVED** [review]
LGTM once this [comment](https://github.com/mlflow/mlflow/pull/13903#discussion_r1867077914) is addressed!

File: mlflow/crewai/__init__.py (hunk 13 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:50]: Interesting, it appeared in my local and the unit tests for this span have passed.
image

File: mlflow/crewai/__init__.py (hunk 14 above)
Comment (TomeHirata) [reply mlflow/crewai/__init__.py:75]: Yeah, I was in two minds. If we implement it in that way, we cannot define `class_method_map` since an AttributeException is raised immediately when the map is defined.

File: mlflow/crewai/__init__.py (hunk 13 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:50]: @TomeHirata My bad, it seems sth weird is happening on my cluster env. Tested it in the new instance and works like a charm😄

File: mlflow/crewai/__init__.py (hunk 14 above)
Comment (B-Step62) [reply mlflow/crewai/__init__.py:75]: What about making `class_method_map` key string and import the class one by one within each try-catch?

Comment (TomeHirata) [reply mlflow/crewai/__init__.py:75]: Let me try that direction!
